contractcourt/htlc_timeout_resolver.go

```diff
@@ -483,13 +489,45 @@ func (h *htlcTimeoutResolver) sweepSecondLevelTx() error {
 			h.broadcastHeight,
 		))
 	}
+
+	// Calculate the budget.
+	//
+	// TODO(yy): the budget is twice the output's value, which is needed as
+	// we don't force sweep the output now. To prevent cascading force
+	// closes, we use all its output value plus a wallet input as the
+	// budget. This is a temporary solution until we can optionally cancel
+	// the incoming HTLC, more details in,
+	// - https://github.com/lightningnetwork/lnd/issues/7969
+	budget := calculateBudget(
+		btcutil.Amount(inp.SignDesc().Output.Value), 2, 0,
+	)
+
+	// For an outgoing HTLC, it must be swept before the RefundTimeout of
+	// its incoming HTLC is reached.
+	//
+	// TODO(yy): we may end up mixing inputs with different time locks.
+	// Suppose we have two outgoing HTLCs,
+	// - HTLC1: nLocktime is 800000, CLTV delta is 80.
+	// - HTLC2: nLocktime is 800001, CLTV delta is 79.
+	// This means they would both have an incoming HTLC that expires at
+	// 800080, hence they share the same deadline but different locktimes.
+	// However, with current design, when we are at block 800000, HTLC1 is
+	// offered to the sweeper. When block 800001 is reached, HTLC1's
+	// sweeping process is already started, while HTLC2 is being offered to
+	// the sweeper, so they won't be mixed. This can become an issue tho,
+	// if we decide to sweep per X blocks. Or the contractcourt sees the
+	// block first while the sweeper is only aware of the last block. To
+	// properly fix it, we need `blockbeat` to make sure subsystems are in
+	// sync.
```

Comment:
In PR 8 we now sweep immediately during startup, so this won't be an issue.

Reply:
That helps, but there's still a race here.  If blocks 800000 and 800001 arrive at the same time (perhaps due to selfish mining), the contractcourt may still offer both HTLCs during the same block (800001), and the sweeper would batch them together.

I think we should fix this properly, instead of relying on the timing of input offerings and block arrivals, which seems fragile.

Reply:
> the contractcourt may still offer both HTLCs during the same block (800001), and the sweeper would batch them together

I think I'm missing something...what's the issue with this scenario?

Reply:
htlcs with different locktimes cannot grouped together. I would also prefer a clean locktime separation so that we don't run in race conditions.

Reply:
> I think I'm missing something...what's the issue with this scenario?

HTLC-Timeout transaction signatures commit to the locktime.  If we try to batch them together when locktimes differ, at least one of the signatures becomes invalid.

And since there's no logic to detect bad inputs, funds are likely to be lost.

Reply:
yup, the locktime logic from the simpleaggregator needs to be kept